mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala

```diff
@@ -0,0 +1,360 @@
+
+  import org.apache.spark.ml.regression.BaseFactorizationMachinesGradient.{LogisticLoss, parseLoss}
+  import org.apache.spark.ml.regression.FMRegressor.initCoefficients
+
+  @Since("3.0.0")
+  def this() = this(Identifiable.randomUID("fmc"))
+
+  /**
+   * Set the dimensionality of the factors.
+   * Default is 8.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setNumFactors(value: Int): this.type = set(numFactors, value)
+  setDefault(numFactors -> 8)
+
+  /**
+   * Set whether to fit global bias term.
+   * Default is true.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setFitBias(value: Boolean): this.type = set(fitBias, value)
+  setDefault(fitBias -> true)
+
+  /**
+   * Set whether to fit linear term.
+   * Default is true.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setFitLinear(value: Boolean): this.type = set(fitLinear, value)
+  setDefault(fitLinear -> true)
+
+  /**
+   * Set the L2 regularization parameter.
+   * Default is 0.0.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setRegParam(value: Double): this.type = set(regParam, value)
+  setDefault(regParam -> 0.0)
+
+  /**
+   * Set the mini-batch fraction parameter.
+   * Default is 1.0.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setMiniBatchFraction(value: Double): this.type = {
+    require(value > 0 && value <= 1.0,
+      s"Fraction for mini-batch SGD must be in range (0, 1] but got $value")
+    set(miniBatchFraction, value)
+  }
+  setDefault(miniBatchFraction -> 1.0)
+
+  /**
+   * Set the standard deviation of initial coefficients.
+   * Default is 0.01.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setInitStd(value: Double): this.type = set(initStd, value)
+  setDefault(initStd -> 0.01)
+
+  /**
+   * Set the maximum number of iterations.
+   * Default is 100.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setMaxIter(value: Int): this.type = set(maxIter, value)
+  setDefault(maxIter -> 100)
+
+  /**
+   * Set the initial step size for the first step (like learning rate).
+   * Default is 1.0.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setStepSize(value: Double): this.type = set(stepSize, value)
+  setDefault(stepSize -> 1.0)
+
+  /**
+   * Set the convergence tolerance of iterations.
+   * Default is 1E-6.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Set the solver algorithm used for optimization.
+   * Default is adamW.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setSolver(value: String): this.type = set(solver, value)
+  setDefault(solver -> AdamW)
+
+  override protected[spark] def train(dataset: Dataset[_]): FMClassifierModel = {
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    train(dataset, handlePersistence)
+  }
+
+  protected[spark] def train(
```

Comment:
I still think that this method is not needed, just create a singe `def train(dataset: Dataset[_]): FMClassifierModel`?

Reply:
> I still think that this method is not needed, just create a singe `def train(dataset: Dataset[_]): FMClassifierModel`?

@zhengruifeng I refer to the implementation of LR. It will handle persistence. If dataset not be persisted, it would persist the dataset, and release the cache after train finish.
Or should I cache dataset regardless of whether the dataset be cached?

LogisticRegression.scala: 481

```scala
  override protected[spark] def train(dataset: Dataset[_]): LogisticRegressionModel = {
    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
    train(dataset, handlePersistence)
  }

  protected[spark] def train(
      dataset: Dataset[_],
      handlePersistence: Boolean): LogisticRegressionModel = instrumented { instr =>
    val instances = extractInstances(dataset)

    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)

    // train model code

    if (handlePersistence) instances.unpersist()
```

Reply:
Yes, but this method `def train(dataset: Dataset[_], handlePersistence: Boolean)` is added for the old [`mllib.LogisticRegression`](https://github.com/apache/spark/blob/6378d4bc06cd1bb1a209bd5fb63d10ef52d75eb4/mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala#L345).

As a new alg, FM do not need to have this method.

>  Or should I cache dataset regardless of whether the dataset be cached?
you should cache the dataset, refer to [LinearSVC](https://github.com/apache/spark/blob/8a4378c6f0ca49f94fdeba08ca095721427e8635/mllib/src/main/scala/org/apache/spark/ml/classification/LinearSVC.scala#L165)

Reply:
```scala
  override protected[spark] def train(dataset: Dataset[_]): FMClassifierModel = {
    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
    val data: RDD[(Double, OldVector)] = ...
    if (handlePersistence) data.persist(StorageLevel.MEMORY_AND_DISK)
    ...
  }
```